libp2p_test.go

```diff
@@ -192,3 +196,52 @@ func TestChainOptions(t *testing.T) {
 		}
 	}
 }
+
+func TestTcpSimultaneousConnect(t *testing.T) {
+	ctx := context.Background()
+
+	// Host1
+	h1, err := New(ctx, Transport(tcp.NewTCPTransport), Security(noise.ID, noise.New),
+		ListenAddrs(ma.StringCast("/ip4/0.0.0.0/tcp/0")))
+	require.NoError(t, err)
+	defer h1.Close()
+
+	// Host2
+	h2, err := New(ctx, Transport(tcp.NewTCPTransport), Security(noise.ID, noise.New),
+		ListenAddrs(ma.StringCast("/ip4/0.0.0.0/tcp/0")))
+	require.NoError(t, err)
+	defer h2.Close()
+
+	h1Info := peer.AddrInfo{
+		ID:    h1.ID(),
+		Addrs: h1.Addrs(),
+	}
+
+	h2Info := peer.AddrInfo{
+		ID:    h2.ID(),
+		Addrs: h2.Addrs(),
+	}
+
+	wg := new(sync.WaitGroup)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for i := 0; i < 20; i++ {
+			require.NoError(t, h1.Connect(context.Background(), h2Info))
+			require.NoError(t, h1.Network().ClosePeer(h2.ID()))
+		}
```

Comment:
Are 20 iterations sufficient to reliably trigger simopen?

Reply:
@raulk Pretty much.

**Here's a run of the same test against `v0.13.0`**(previous release, did not have hole punching), and it fails repeatedly:

```
aarshshah@Aarshs-MacBook-Pro go-libp2p % go test -run TestTcpSimultaneousConnect -count 10
--- FAIL: TestTcpSimultaneousConnect (0.28s)
    libp2p_test.go:231: 
        	Error Trace:	libp2p_test.go:231
        	            				asm_amd64.s:1374
        	Error:      	Received unexpected error:
        	            	failed to dial QmZsP5W7ceKhynLcEtQpHXRTNTfd7i9YnBsnndwvmi5p3t: all dials failed
        	            	  * [/ip4/192.168.43.87/tcp/50093] failed to negotiate security protocol: error reading handshake message: noise: message is too short
        	            	  * [/ip4/127.0.0.1/tcp/50093] failed to negotiate security protocol: error reading handshake message: noise: message is too short
        	Test:       	TestTcpSimultaneousConnect
    libp2p_test.go:241: 
        	Error Trace:	libp2p_test.go:241
        	            				asm_amd64.s:1374
        	Error:      	Received unexpected error:
        	            	failed to dial QmcMLKDi4aSWSkMk1W7pHY5uFiB6Wi7yKVzgVTY6oWL4an: all dials failed
        	            	  * [/ip4/192.168.43.87/tcp/50092] failed to negotiate security protocol: error reading handshake message: noise: message is too short
        	            	  * [/ip4/127.0.0.1/tcp/50092] failed to negotiate security protocol: error reading handshake message: noise: message is too short
        	Test:       	TestTcpSimultaneousConnect
--- FAIL: TestTcpSimultaneousConnect (0.23s)
    libp2p_test.go:241: 
        	Error Trace:	libp2p_test.go:241
        	            				asm_amd64.s:1374
        	Error:      	Received unexpected error:
        	            	failed to dial Qmed9XqtjKjxEEoJi12PwKAf2NW35qkA6L2jffMEaodGQ2: all dials failed
        	            	  * [/ip4/192.168.43.87/tcp/50095] failed to negotiate security protocol: error reading handshake message: noise: message is too short
        	            	  * [/ip4/127.0.0.1/tcp/50095] failed to negotiate security protocol: error reading handshake message: noise: message is too short
        	Test:       	TestTcpSimultaneousConnect
    libp2p_test.go:231: 
        	Error Trace:	libp2p_test.go:231
        	            				asm
        	            				
  ...........................
  ...........................
  
  FAIL
exit status 1
FAIL	github.com/libp2p/go-libp2p	3.496s      	            				
```


**Now, here's the same run with simultaneous connect feature:**

```
aarshshah@Aarshs-MacBook-Pro go-libp2p % go test -run TestTcpSimultaneousConnect -count 10
go: downloading github.com/libp2p/go-libp2p-core v0.8.2
go: downloading github.com/libp2p/go-libp2p-transport-upgrader v0.4.2
go: downloading github.com/libp2p/go-libp2p-swarm v0.4.2
PASS
ok  	github.com/libp2p/go-libp2p	5.609s

```